src/CBot/CBotClass.cpp

```diff
@@ -561,29 +562,27 @@ bool CBotClass::CompileDefItem(CBotToken* &p, CBotCStack* pStack, bool bSecond)
             while ( IsOfType( p, ID_OPBRK ) )   // a table?
             {
                 CBotInstr* i = nullptr;
-
+                pStack->SetStartError( p->GetStart() );
                 if ( p->GetType() != ID_CLBRK )
+                {
                     i = CBotExpression::Compile( p, pStack );           // expression for the value
+                    if (i == nullptr || pStack->GetType() >= CBotTypBoolean) // must be a number
```

Comment:
This is not really a correct fix. Not everything that is a number is a valid index, only integers are.
```C++
float a[];
a[2.3] = 7.5;
a[true] = 6; // this throws a runtime error
```